### ISSUE TITLE
Terraform (NS1 provider) crashes on inconsistent list element

### ISSUE THREAD

**Issue by neilstuartcraig** [hashicorp/terraform#25936]
I was trying to use a list as a variable default in a module, using the NS1 provider and didn't realise that all the child items of the list have to be consistent, my var def looks like this:

```terraform
variable "record_filters" {
    type = list(object(
    {
        filter = string,
        config = map(any)
    }));
    default = [
        {
            "filter" = "up",
            "config" = {}
        },
        {
            "filter" = "shuffle",
            "config" = {},
        },
        {
            "filter" = "select_first_n",
            "config" = {
                N = 2
            }
        }
    ]

    description = "A definition of the filters to apply to this record. Filters will be executed in the stated order"
}
```

I now know this is wrong but it crashed Terraform which seems suboptimal (I think). Error:

```
panic: inconsistent list element types (cty.Object(map[string]cty.Type{"config":cty.Map(cty.DynamicPseudoType), "filter":cty.String}) then cty.Object(map[string]cty.Type{"config":cty.Map(cty.Number), "filter":cty.String}))
```

### Terraform Version
```
Terraform v0.12.26
+ provider.ns1 v1.8.4
```

### Terraform Configuration Files
```terraform
# DNS records which automatically switch datacentres in/out depending on a status indicator for each datacentre

variable "zone" {
    type = string
    description = "The environment/workspace-specific dynamic DNS record - dynamic.bbc.co.uk and staging equiv"
}

variable "record" {
    type = string
    description = "The DNS record/domain e.g. www.bbc.co.uk"
}

variable "record_type" {

# TODO: change this to type of `set`, see https://www.terraform.io/docs/configuration/variables.html#type-constraints
    type = string
    description = "The DNS record type, e.g. A, AAAA, CNAME, TXT etc."
}


variable "ttl" {
    type = number  
    description = "The DNS record TTL, in seconds"
    default = 300
}

variable "records" {
    type = map(object({
        datacentre = string
        answer = string
    }))
    # No default because we always need to have at least 1 record

# TODO: add link to docs
    description = "A definition of the regions to use in this configuration <INSERT LINK TO DOCS>"
}

variable "record_filters" {
    type = list(object(
    {
        filter = string,
        config = map(any)
    }));
    default = [
        {
            "filter" = "up",
            "config" = {}
        },
        {
            "filter" = "shuffle",
            "config" = {},
        },
        {
            "filter" = "select_first_n",
            "config" = {
                N = 2
            }
        }
    ]

    description = "A definition of the filters to apply to this record. Filters will be executed in the stated order"
}

variable datacentre_statuses {
    type = map
    description = "Status (boolean) for each datacentre to define whether they're currently up or down"
}


resource "ns1_record" "dc_failover_record" {
    zone   = var.zone
    domain = var.record
    use_client_subnet = true
    type   = var.record_type
    ttl    = var.ttl


    dynamic "answers" {
        for_each = var.records
        content {
            answer = answers.value["answer"]
            meta = {
                up = var.datacentre_statuses[answers.value["datacentre"]]            
                note = "DC: ${answers.value["datacentre"]}"
                note = "VIP: ${answers.key}"
            }
        }
    }

    dynamic "filters" {
        for_each = var.record_filters
        content {
            filter = filters.value["filter"]
            config = filters.value["config"]
        }
    }
}
```

### Crash Output
```
2020/08/20 14:04:08 [INFO] Terraform version: 0.12.26  
2020/08/20 14:04:08 [INFO] Go runtime version: go1.13.11
2020/08/20 14:04:08 [INFO] CLI args: []string{"/usr/local/bin/terraform", "apply"}
2020/08/20 14:04:08 [DEBUG] Attempting to open CLI config file: /Users/craign04/.terraformrc
2020/08/20 14:04:08 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2020/08/20 14:04:08 [INFO] CLI command args: []string{"apply"}
2020/08/20 14:04:08 [TRACE] Meta.Backend: no config given or present on disk, so returning nil config
2020/08/20 14:04:08 [TRACE] Meta.Backend: backend has not previously been initialized in this working directory
2020/08/20 14:04:08 [DEBUG] New state was assigned lineage "39854b5a-44cb-4742-8c0b-5396e1147c24"
2020/08/20 14:04:08 [TRACE] Meta.Backend: using default local state only (no backend configuration, and no existing initialized backend)
2020/08/20 14:04:08 [TRACE] Meta.Backend: instantiated backend of type <nil>
2020/08/20 14:04:08 [DEBUG] checking for provider in "."
2020/08/20 14:04:08 [DEBUG] checking for provider in "/usr/local/bin"
2020/08/20 14:04:08 [DEBUG] checking for provider in ".terraform/plugins/darwin_amd64"
2020/08/20 14:04:08 [DEBUG] found provider "terraform-provider-ns1_v1.8.4_x4"
2020/08/20 14:04:08 [DEBUG] found valid plugin: "ns1", "1.8.4", "/Users/craign04/dev/ns1-dns-config/zones/primaries/.terraform/plugins/darwin_amd64/terraform-provider-ns1_v1.8.4_x4"
2020/08/20 14:04:08 [DEBUG] checking for provisioner in "."
2020/08/20 14:04:08 [DEBUG] checking for provisioner in "/usr/local/bin"
2020/08/20 14:04:08 [DEBUG] checking for provisioner in ".terraform/plugins/darwin_amd64"
2020/08/20 14:04:08 [TRACE] Meta.Backend: backend <nil> does not support operations, so wrapping it in a local backend
2020/08/20 14:04:08 [INFO] backend/local: starting Apply operation
2020/08/20 14:04:08 [TRACE] backend/local: requesting state manager for workspace "staging"
2020/08/20 14:04:08 [TRACE] backend/local: state manager for workspace "staging" will:
 - read initial snapshot from terraform.tfstate.d/staging/terraform.tfstate
 - write new snapshots to terraform.tfstate.d/staging/terraform.tfstate
 - create any backup at terraform.tfstate.d/staging/terraform.tfstate.backup
2020/08/20 14:04:08 [TRACE] backend/local: requesting state lock for workspace "staging"
2020/08/20 14:04:08 [TRACE] statemgr.Filesystem: preparing to manage state snapshots at terraform.tfstate.d/staging/terraform.tfstate
2020/08/20 14:04:08 [TRACE] statemgr.Filesystem: existing snapshot has lineage "ad4112f0-991c-b2d2-1626-efa96e17154d" serial 314
2020/08/20 14:04:08 [TRACE] statemgr.Filesystem: locking terraform.tfstate.d/staging/terraform.tfstate using fcntl flock
2020/08/20 14:04:08 [TRACE] statemgr.Filesystem: writing lock metadata to terraform.tfstate.d/staging/.terraform.tfstate.lock.info
2020/08/20 14:04:08 [TRACE] backend/local: reading remote state for workspace "staging"
2020/08/20 14:04:08 [TRACE] statemgr.Filesystem: reading latest snapshot from terraform.tfstate.d/staging/terraform.tfstate
2020/08/20 14:04:08 [TRACE] statemgr.Filesystem: read snapshot with lineage "ad4112f0-991c-b2d2-1626-efa96e17154d" serial 314
2020/08/20 14:04:08 [TRACE] backend/local: retrieving local state snapshot for workspace "staging"
2020/08/20 14:04:08 [TRACE] backend/local: building context for current working directory
2020/08/20 14:04:08 [TRACE] statemgr.Filesystem: removing lock metadata file terraform.tfstate.d/staging/.terraform.tfstate.lock.info
2020/08/20 14:04:08 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate.d/staging/terraform.tfstate using fcntl flock
panic: inconsistent list element types (cty.Object(map[string]cty.Type{"config":cty.Map(cty.DynamicPseudoType), "filter":cty.String}) then cty.Object(map[string]cty.Type{"config":cty.Map(cty.Number), "filter":cty.String}))

goroutine 51 [running]:
github.com/zclconf/go-cty/cty.ListVal(0xc00016c720, 0x3, 0x3, 0xc00059b290, 0xc000488fb0, 0x1, 0x1)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/pkg/mod/github.com/zclconf/go-cty@v1.2.1/cty/value_init.go:166 +0x436
github.com/zclconf/go-cty/cty/convert.conversionTupleToList.func2(0x2fd4d60, 0xc0004953c0, 0x269ffe0, 0xc0004953e0, 0xc000488fb0, 0x1, 0x1, 0xc00054a000, 0x10, 0x10, ...)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/pkg/mod/github.com/zclconf/go-cty@v1.2.1/cty/convert/conversion_collection.go:267 +0x53f
github.com/zclconf/go-cty/cty/convert.getConversion.func1(0x2fd4d60, 0xc0004953c0, 0x269ffe0, 0xc0004953e0, 0x0, 0x0, 0x0, 0xc00059b620, 0xc000488fa0, 0x2fd4c60, ...)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/pkg/mod/github.com/zclconf/go-cty@v1.2.1/cty/convert/conversion.go:46 +0x1ab
github.com/zclconf/go-cty/cty/convert.retConversion.func1(0x2fd4d60, 0xc0004953c0, 0x269ffe0, 0xc0004953e0, 0xc000488fa0, 0x0, 0x0, 0x3, 0x2, 0x0)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/pkg/mod/github.com/zclconf/go-cty@v1.2.1/cty/convert/conversion.go:179 +0x6b
github.com/zclconf/go-cty/cty/convert.Convert(0x2fd4d60, 0xc0004953c0, 0x269ffe0, 0xc0004953e0, 0x2fd4c60, 0xc000488d90, 0x0, 0x0, 0x0, 0xc00059a5d0, ...)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/pkg/mod/github.com/zclconf/go-cty@v1.2.1/cty/convert/public.go:51 +0x1aa
github.com/hashicorp/terraform/configs.decodeVariableBlock(0xc00058e680, 0x0, 0x2adab31, 0x8, 0x1, 0xc0005a9440)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/named_values.go:109 +0x8c0
github.com/hashicorp/terraform/configs.(*Parser).loadConfigFile(0xc00059c740, 0xc000236b20, 0x1b, 0x0, 0x0, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/parser_config.go:96 +0x1c24
github.com/hashicorp/terraform/configs.(*Parser).LoadConfigFile(...)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/parser_config.go:22
github.com/hashicorp/terraform/configs.(*Parser).loadFiles(0xc00059c740, 0xc00010daf0, 0x1, 0x1, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/parser_config_dir.go:77 +0x342
github.com/hashicorp/terraform/configs.(*Parser).LoadConfigDir(0xc00059c740, 0xc000040680, 0x13, 0x1, 0xc0001a24a8, 0xc000043301, 0x10)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/parser_config_dir.go:37 +0x121
github.com/hashicorp/terraform/configs/configload.(*Loader).moduleWalkerLoad(0xc0000b6820, 0xc00059e120, 0x0, 0x0, 0x0, 0xc0001e9350, 0x101742e)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/configload/loader_load.go:87 +0x2c4
github.com/hashicorp/terraform/configs/configload.(*Loader).makeModuleWalkerSnapshot.func1(0xc00059e120, 0x100c688, 0x120, 0x29a3480, 0x1, 0xc00059e120)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/configload/loader_snapshot.go:134 +0x60
github.com/hashicorp/terraform/configs.ModuleWalkerFunc.LoadModule(0xc000494340, 0xc00059e120, 0x1, 0x1, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/config_build.go:111 +0x30
github.com/hashicorp/terraform/configs.buildChildModules(0xc000563c70, 0x2f92ce0, 0xc000494340, 0xc000494340, 0xc0004f5750, 0x1bc9531, 0x289a8e0)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/config_build.go:58 +0x52a
github.com/hashicorp/terraform/configs.BuildConfig(0xc00048cd00, 0x2f92ce0, 0xc000494340, 0xc000494340, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/config_build.go:24 +0x71
github.com/hashicorp/terraform/configs/configload.(*Loader).LoadConfigWithSnapshot(0xc0000b6820, 0x2ad30cd, 0x1, 0x10ecd9a, 0x0, 0x0, 0xc0004f5c08, 0xc0000b6000)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/configs/configload/loader_snapshot.go:31 +0x10d
github.com/hashicorp/terraform/backend/local.(*Local).contextDirect(0xc000244000, 0xc0001cc700, 0x0, 0x0, 0xc00048a020, 0x0, 0x0, 0x0, 0x0, 0xc00010dda0, ...)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/backend/local/backend_local.go:132 +0x56
github.com/hashicorp/terraform/backend/local.(*Local).context(0xc000244000, 0xc0001cc700, 0x1, 0x1, 0x3, 0xc00027b5f0, 0x1, 0x3, 0xc000120310)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/backend/local/backend_local.go:94 +0xd27
github.com/hashicorp/terraform/backend/local.(*Local).opApply(0xc000244000, 0x2fd30e0, 0xc0005a8ec0, 0x2fd30e0, 0xc0005a8f00, 0xc0001cc700, 0xc0005a8e80)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/backend/local/backend_apply.go:53 +0x302
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc00010de60, 0xc00010de70, 0xc00010de80, 0xc0001cc700, 0xc000244000, 0xc0005a8e80, 0xc00010de50, 0x2fd30e0, 0xc0005a8ec0, 0x2fd30e0, ...)
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/backend/local/backend.go:350 +0x1c0
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
	/private/tmp/terraform-20200527-43164-gdmmt5/terraform-0.12.26/src/github.com/hashicorp/terraform/backend/local/backend.go:334 +0x366
```

### Expected Behavior
I expected an error to be returned

### Actual Behavior
Terraform crashed

### Steps to Reproduce
1. `terraform init`
2. `terraform apply`

**Comment by neilstuartcraig** [comment]
Just noticed all the typos in my config (the variable def). I wonder if those are not caught?

**Comment by alisdair** [comment]
Thank you for the detailed report! I was able to reproduce the crash on 0.12.28.

This seems to be a duplicate of #21588, which has now been fixed in the upstream library and was released as part of 0.13.0. I've confirmed that the bug is fixed in the 0.13 series.
